python/lib/client/dmod/client/dmod_client.py

```diff
@@ -1,277 +1,213 @@
 
 
 class DmodClient:
 
-    def __init__(self, client_config: YamlClientConfig, bypass_request_service: bool = False, *args, **kwargs):
+    def __init__(self, client_config: ClientConfig, bypass_request_service: bool = False, *args, **kwargs):
         self._client_config = client_config
-        self._dataset_client = None
-        self._ngen_client = None
-        self._ngen_cal_client = None
+        self._data_service_client = None
+        self._job_client = None
         self._bypass_request_service = bypass_request_service
 
-    @property
-    def client_config(self):
-        return self._client_config
-
-    async def create_dataset(self, dataset_name: str, category: DataCategory, domain: Optional[DataDomain] = None,
-                             **kwargs) -> bool:
-        """
-        Create a dataset from the given parameters.
+        # TODO: this should (optionally) be a client multiplexer (once that is available) instead of a transport client
+        #  (with a getter to actually get a transport client in either case)
+        request_t_client_type = determine_transport_client_type(client_config.request_service.endpoint_protocol,
+                                                                WebSocketClient)
+        self._request_service_conn: TransportLayerClient = request_t_client_type(**client_config.request_service.dict())
 
-        Note that despite the type hinting, ``domain`` is only semi-optional, as a domain is required to create a
-        dataset. However, if a ``data_format`` keyword arg provides a ::class:`DataFormat` value, then a minimal
-        ::class:`DataDomain` object can be generated and used.
+        self._auth_client: AuthClient = AuthClient(transport_client=self._get_transport_client())
 
-        Additionally, ``continuous_restrictions`` and ``discrete_restrictions`` keyword args are used if present for
-        creating the domain when necessary.  If neither are provided, the generated domain will have a minimal discrete
-        restriction created for "all values" (i.e., an empty list) of the first index variable of the provided
-        ::class:`DataFormat`.
+    def _get_transport_client(self, **kwargs) -> TransportLayerClient:
+        # TODO: later add support for multiplexing capabilities and spawning wrapper clients
+        return self._request_service_conn
 
-        In the event neither a domain not a data format is provided, a ::class:`ValueError` is raised.
+    @property
+    def client_config(self) -> ClientConfig:
+        return self._client_config
 
-        Additionally, keyword arguments are forwarded in the call to the ::attribute:`dataset_client` property's
-        ::method:`DatasetClient.create_dataset` function.  This includes the aforementioned kwargs for a creating a
-        default ::class:`DataDomain`, but only if they are otherwise ignored because a valid domain arg was provided.
+    async def data_service_action(self, action: str, **kwargs) -> ResultIndicator:
```

Comment:
I'm not sure it will be as valuable as it might first appear, because of the way the class is used, but it's trivial enough that I'm ok with adding it.

Reply:
Thanks for doing some digging. Determining what is supported in x version of python is always tedious. The general community guidance so far is to use [`typing_extensions`](https://github.com/python/typing_extensions). This is a library that ports `python.typing` features found in later version of python. We are already using it in the client, communication, core, and scheduler libraries to extend support. I would recommend that we use it in this case too.

The python 3.12 question is a much larger one and IMO a can of worms. `3.12` introduces a new `type` keyword that lets you effectively create what are now `TypeAlias`. However, until `3.12` is our minimum version we will need either `typing.TypeAlias` or `typing_extensions.TypeAlias`. So, to me, the 3.12 question it is worth punting on for now. Thoughts?

Reply:
Let's leave this out for now.  It's non-essential, and it adds eventual technical debt with an obfuscated timeline.  On top of that, it doesn't cover the entire scope of where these strings are relevant; i.e., it may leave a trap to fall into later when CLI arg parsing needs some changes.

We can still consider it, but let's look at that separately.